src/zeroconf/_services/info.py

```diff
@@ -355,57 +367,75 @@ def async_update_records(self, zc: 'Zeroconf', now: float, records: List[RecordU
+
+    def _process_records_threadsafe(self, zc: 'Zeroconf', now: float, records: List[RecordUpdate]) -> bool:
+        """Thread safe record updating.
 
-    def _process_records_threadsafe(self, zc: 'Zeroconf', now: float, records: List[RecordUpdate]) -> None:
-        """Thread safe record updating."""
+        Returns True if new records were added.
+        """
         seen_addresses: Set[bytes] = set()
+        updated: bool = False
         for record_update in records:
             record = record_update.new
             if isinstance(record, DNSAddress):
                 seen_addresses.add(record.address)
-            self._process_record_threadsafe(record, now)
+            updated |= self._process_record_threadsafe(record, now)
         for record in self._get_address_records_from_cache(zc):
             if record.address not in seen_addresses:
-                self._process_record_threadsafe(record, now)
+                updated |= self._process_record_threadsafe(record, now)
+        return updated
+
+    def _process_record_threadsafe(self, record: DNSRecord, now: float) -> bool:
+        """Thread safe record updating.
 
-    def _process_record_threadsafe(self, record: DNSRecord, now: float) -> None:
-        """Thread safe record updating."""
+        Returns True if a new record was added.
+        """
         if record.is_expired(now):
-            return
+            return False
 
-        if isinstance(record, DNSAddress):
-            if record.key != self.server_key:
-                return
+        if record.key == self.server_key and isinstance(record, DNSAddress):
             try:
                 ip_addr = _cached_ip_addresses(record.address)
             except ValueError as ex:
                 log.warning("Encountered invalid address while processing %s: %s", record, ex)
-                return
-            if isinstance(ip_addr, ipaddress.IPv4Address):
-                if ip_addr not in self._ipv4_addresses:
-                    self._ipv4_addresses.insert(0, ip_addr)
-                return
-            if ip_addr not in self._ipv6_addresses:
-                self._ipv6_addresses.insert(0, ip_addr)
-                if ip_addr.is_link_local:
-                    self.interface_index = record.scope_id
-            return
+                return False
+
+            target_list: Union[List[ipaddress.IPv4Address], List[ipaddress.IPv6Address]] = (
+                self._ipv4_addresses if ip_addr.version == 4 else self._ipv6_addresses
+            )
+            ip_in_list = ip_addr in target_list
+
+            if not ip_in_list:
+                # Add the address to the front of the list
+                target_list.insert(0, ip_addr)
+                return True
+            elif ip_in_list and target_list and target_list[0] != ip_addr:
+                # Move the address to the front of the list
+                target_list.pop(target_list.index(ip_addr))
+                target_list.insert(0, ip_addr)
```

Comment:
This might be unreachable since as soon as we have an address its considered complete